{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2e716cd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.mode.chained_assignment = None\n",
    "import sys\n",
    "import copy\n",
    "import cobra\n",
    "import cobra.test\n",
    "import math\n",
    "from collections import defaultdict\n",
    "import efmtool\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbad63b7",
   "metadata": {},
   "source": [
    "### Define functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a7859efb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_to_irreversible(cobra_model):\n",
    "    \"\"\"\n",
    "    Split reversible reactions into two irreversible reactions\n",
    "\n",
    "    These two reactions will proceed in opposite directions. This\n",
    "    guarentees that all reactions in the model will only allow\n",
    "    positive flux values, which is useful for some modeling problems.\n",
    "\n",
    "    cobra_model:\n",
    "        A cobra model object which will be modified in place.\n",
    "\n",
    "    \"\"\"\n",
    "    from cobra import Reaction\n",
    "    import pandas as pd\n",
    "    from cobra.util import set_objective\n",
    "    reactions_to_add = []\n",
    "    coefficients = {}\n",
    "    for reaction in cobra_model.reactions:\n",
    "        # If a reaction is reverse only, the forward reaction (which\n",
    "        # will be constrained to 0) will be left in the model.\n",
    "        if reaction.lower_bound < 0:\n",
    "            reverse_reaction = Reaction(reaction.id + \"_b\")\n",
    "            reaction.id = reaction.id + \"_f\"\n",
    "            reverse_reaction.lower_bound = max(0, -reaction.upper_bound)\n",
    "            reverse_reaction.upper_bound = -reaction.lower_bound\n",
    "            coefficients[reverse_reaction] = reaction.objective_coefficient * -1\n",
    "            reaction.upper_bound = max(0, reaction.upper_bound)\n",
    "            reaction.lower_bound = max(0, reaction.lower_bound)\n",
    "\n",
    "            # Make the directions aware of each other\n",
    "            reaction.notes[\"reflection\"] = reverse_reaction.id\n",
    "            reverse_reaction.notes[\"reflection\"] = reaction.id\n",
    "            reaction_dict = {k: v * -1 for k, v in pd.Series.iteritems(reaction._metabolites)}\n",
    "            reverse_reaction.add_metabolites(reaction_dict)\n",
    "            reverse_reaction._model = reaction._model\n",
    "            reverse_reaction._genes = reaction._genes\n",
    "            for gene in reaction._genes:\n",
    "                gene._reaction.add(reverse_reaction)\n",
    "            reverse_reaction.subsystem = reaction.subsystem\n",
    "            #reverse_reaction._gene_reaction_rule = reaction._gene_reaction_rule\n",
    "            #\n",
    "            # ALO edited previous line as\n",
    "            #\n",
    "            reverse_reaction._gene_reaction_rule = reaction.gene_reaction_rule\n",
    "            reactions_to_add.append(reverse_reaction)\n",
    "    cobra_model.add_reactions(reactions_to_add)\n",
    "    set_objective(cobra_model, coefficients, additive=True)\n",
    "    return cobra_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fd9ce169",
   "metadata": {},
   "outputs": [],
   "source": [
    "def MinNW(model,protected_mets,protected_rxns,functionality_df):\n",
    "    '''\n",
    "    A MILP formulation derived from (https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-016-1412-z) to\n",
    "    reduce a Genome-Scale Metabolic Network so that EFM calculation is faster and macroreactions may be derived. This is \n",
    "    an approach that InSilico uses. I want to make sure that it works for my stuff as well. \n",
    "    This is the second MILP, MinNW, that finds the smallest network it can given a model and specific functionalities. \n",
    "    \n",
    "    Inputs:\n",
    "        model: A cobra genome-scale metabolic model\n",
    "        protected_mets: A list of protected metabolites, i.e. metabolites that absolutely need to be in the resulting network\n",
    "        \n",
    "        protected_rxns: A list of protected reactions, i.e. reactions that absolutely need to be in the resulting network\n",
    "        \n",
    "        functionality_df: A (Fx3) pandas DataFrame containing the reactions along with their lower and upper bounds describing \n",
    "            the F functionalities desired in the resulting network. May be opposing ones (e.g. O2 uptake inhibited and allowed).\n",
    "            The first column contains the name of the reactions of interest, the second contains the lower bound and third\n",
    "            the upper bound for the reaction. May contain more than one reaction. Needs to contain an equivalent amount of bonds. \n",
    "        \n",
    "    Outputs:\n",
    "        prob.status: Status of solver, either 'infeasible' or 'optimal'\n",
    "        \n",
    "        model_dict: A dictionary containing the Stoichiometric matrix, reactions, metabolites and bounds for the new, reduced \n",
    "            network. \n",
    "    \n",
    "    '''\n",
    "    import cvxpy as cp\n",
    "    import cvxopt\n",
    "    from cvxopt import solvers\n",
    "    import cobra\n",
    "    import itertools\n",
    "    \n",
    "    A = cobra.util.array.create_stoichiometric_matrix(model,'dense')\n",
    "    b = np.zeros(A.shape[0])\n",
    "    rxns = [rxn.id for rxn in model.reactions]\n",
    "    n_rxns = len(rxns)\n",
    "    n_mets = A.shape[0]\n",
    "    n_functions = functionality_df.shape[0]\n",
    "    lb = [rxn.lower_bound for rxn in model.reactions]\n",
    "    ub = [rxn.upper_bound for rxn in model.reactions]\n",
    "    #M = max(ub)\n",
    "    M = np.array([ub]).T\n",
    "    #print(M.shape)\n",
    "    delta = 1e-6 # Between 1e-4 and 1e-6\n",
    "    #delta = np.array([lb]).T # Between 1e-4 and 1e-6\n",
    "    \n",
    "    # Initiate an id_protected_rxns list:\n",
    "    protected_rxns_orig = list(itertools.chain.from_iterable(protected_rxns))\n",
    "    protected_met_rxns = []\n",
    "    for met in protected_mets:\n",
    "        protected_met_rxns_tmp = find_rxns_from_mets(model,met)\n",
    "        if bool(set(protected_rxns_orig) & set(protected_met_rxns_tmp)):\n",
    "            continue\n",
    "        else:\n",
    "            protected_met_rxns.append(protected_met_rxns_tmp)\n",
    "    protected_met_rxns = list(set(list(itertools.chain.from_iterable(protected_met_rxns))))\n",
    "    \n",
    "    # Set constraints so that reactions associated with protected metabolites have at least \n",
    "    # an alpha of 1:\n",
    "    id_met_rxn_protected = []\n",
    "    for rxn in protected_met_rxns:\n",
    "        id_met_rxn_protected.append(rxns.index(rxn))\n",
    "    prot_met_vec = np.zeros(n_rxns)\n",
    "    prot_met_vec[np.array(id_met_rxn_protected)] = 1\n",
    "    \n",
    "    id_rxn_protected = []\n",
    "    for rxn in protected_rxns:\n",
    "        id_rxn_protected.append(rxns.index(rxn))\n",
    "    prot_rxn_vec = np.zeros(n_rxns)\n",
    "    prot_rxn_vec[np.array(id_rxn_protected)] = 1\n",
    "    \n",
    "    # Add constraints to a matrix instead of multiple dictionaries:\n",
    "    alpha_output = cp.Variable(n_rxns,boolean=True)\n",
    "    alpha_orig = cp.Variable((n_rxns,n_functions),boolean=True) # Create a matrix of unknown alphas.\n",
    "    v_vector_orig = cp.Variable((n_rxns,n_functions)) # Create a matrix of unknown fluxes.\n",
    "    \n",
    "    # Set upper and lower matrices:\n",
    "    lb_mat = np.array(([lb]*n_functions)).T\n",
    "    ub_mat = np.array(([ub]*n_functions)).T\n",
    "    \n",
    "    # Set bounds according to input df:\n",
    "    for j in range(0,n_functions):\n",
    "        # Get id of reactions in function first column:\n",
    "        id_vec_tmp = [rxns.index(x) for x in functionality_df.iloc[j,0]]\n",
    "        lb_mat[id_vec_tmp,j] = functionality_df.iloc[j,1]\n",
    "        ub_mat[id_vec_tmp,j] = functionality_df.iloc[j,2]\n",
    "    \n",
    "    # Set the steady-state conditions:\n",
    "    b_mat = np.zeros((n_functions,n_mets)).T\n",
    "    \n",
    "    # Set the inequalities:\n",
    "    my_constraints = []\n",
    "    my_constraints.append(A @ v_vector_orig == b_mat)\n",
    "    my_constraints.append(v_vector_orig >= lb_mat)\n",
    "    my_constraints.append(v_vector_orig <= ub_mat)\n",
    "    my_constraints.append(delta*alpha_orig <= v_vector_orig)\n",
    "    #my_constraints.append(0 <= v_vector_orig)\n",
    "    my_constraints.append(v_vector_orig <= cp.multiply(alpha_orig,M))\n",
    "    #my_constraints.append(v_vector_orig <= alpha_orig*M)\n",
    "    \n",
    "    # Add the protected reaction constraints:\n",
    "    my_constraints.append(cp.sum(cp.multiply(prot_rxn_vec,alpha_output)) == len(id_rxn_protected))\n",
    "    \n",
    "    # Add the protected metabolite constraints:\n",
    "    my_constraints.append(cp.sum(cp.multiply(prot_met_vec,alpha_output)) >= 1)\n",
    "    \n",
    "    # Calculate the sum of alphas:\n",
    "    alpha_sum = cp.sum(alpha_orig, axis=1, keepdims=False)\n",
    "    my_constraints.append(alpha_output <= alpha_sum)\n",
    "    my_constraints.append(alpha_sum <= (1+n_functions)*alpha_output)\n",
    "    \n",
    "    # Solve!\n",
    "    prob = cp.Problem(cp.Minimize(cp.sum(alpha_output)),constraints=my_constraints)\n",
    "    #prob.solve(solver = 'GLPK_MI')\n",
    "    prob.solve(solver = 'CPLEX')\n",
    "    \n",
    "    # Output a new model (S-matrix):\n",
    "    id_present_reactions = np.where(alpha_output.value)\n",
    "    A2 = A.T[id_present_reactions].T\n",
    "    id_metabolites_present = np.where(A2.any(axis=1))[0]\n",
    "    A_new = np.delete(A2, np.where(~A2.any(axis=1))[0], 0)\n",
    "    \n",
    "    # Output new lbs, ubs, rxns and metabolites for the model:\n",
    "    lb_new = list(np.array(lb)[id_present_reactions])\n",
    "    ub_new = list(np.array(ub)[id_present_reactions])\n",
    "    rxns_new = list(np.array(rxns)[id_present_reactions])\n",
    "    metabolites = [met.id for met in model.metabolites]\n",
    "    metabolites_new = list(np.array(metabolites)[id_metabolites_present])\n",
    "    \n",
    "    # Output a model as dictionary:\n",
    "    model_dict = defaultdict()\n",
    "    model_dict['A'] = A_new\n",
    "    model_dict['rxns'] = rxns_new\n",
    "    model_dict['mets'] = metabolites_new\n",
    "    model_dict['ub'] = ub_new\n",
    "    model_dict['lb'] = lb_new\n",
    "    \n",
    "    return prob.status,model_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c0b670bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def MinNW_iterative(model,protected_mets,protected_rxns,functionality_df):\n",
    "    '''\n",
    "    An iterative MILP formulation derived from (https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-016-1412-z) \n",
    "    to reduce a Genome-Scale Metabolic Network so that EFM calculation is faster and macroreactions may be derived. This is \n",
    "    an approach that InSilico uses. I want to make sure that it works for my stuff as well. \n",
    "    This is the third MILP, MinNW_iterative, that finds all the smallest network it can given a model and specific \n",
    "    functionalities. It goes through the process iteratively until all the smallest subnetworks are found. \n",
    "    \n",
    "    Inputs:\n",
    "        model: A cobra genome-scale metabolic model\n",
    "        protected_mets: A list of protected metabolites, i.e. metabolites that absolutely need to be in the resulting network\n",
    "        \n",
    "        protected_rxns: A list of protected reactions, i.e. reactions that absolutely need to be in the resulting network\n",
    "        \n",
    "        functionality_df: A (Fx3) pandas DataFrame containing the reactions along with their lower and upper bounds describing \n",
    "            the F functionalities desired in the resulting network. May be opposing ones (e.g. O2 uptake inhibited and allowed).\n",
    "            The first column contains the name of the reactions of interest, the second contains the lower bound and third\n",
    "            the upper bound for the reaction. May contain more than one reaction. Needs to contain an equivalent amount of bonds. \n",
    "        \n",
    "    Outputs:\n",
    "        alpha_matrix: A (n_rxns x n_subnetworks) boolean matrix indicating which reactions are present in all the \n",
    "        resulting reduced networks. \n",
    "    \n",
    "    '''\n",
    "    import cvxpy as cp\n",
    "    import cvxopt\n",
    "    import cobra\n",
    "    import itertools\n",
    "    \n",
    "    A = cobra.util.array.create_stoichiometric_matrix(model,'dense')\n",
    "    b = np.zeros(A.shape[0])\n",
    "    rxns = [rxn.id for rxn in model.reactions]\n",
    "    n_rxns = len(rxns)\n",
    "    n_mets = A.shape[0]\n",
    "    n_functions = functionality_df.shape[0]\n",
    "    lb = [rxn.lower_bound for rxn in model.reactions]\n",
    "    ub = [rxn.upper_bound for rxn in model.reactions]\n",
    "    #M = max(ub)\n",
    "    M = np.array([ub]).T\n",
    "    #print(M.shape)\n",
    "    delta = 1e-6 # Between 1e-4 and 1e-6\n",
    "    #delta = np.array([lb]).T # Between 1e-4 and 1e-6\n",
    "    \n",
    "    # Initiate an id_protected_rxns list:\n",
    "    protected_rxns_orig = list(itertools.chain.from_iterable(protected_rxns))\n",
    "    protected_met_rxns = []\n",
    "    for met in protected_mets:\n",
    "        protected_met_rxns_tmp = find_rxns_from_mets(model,met)\n",
    "        if bool(set(protected_rxns_orig) & set(protected_met_rxns_tmp)):\n",
    "            continue\n",
    "        else:\n",
    "            protected_met_rxns.append(protected_met_rxns_tmp)\n",
    "    protected_met_rxns = list(set(list(itertools.chain.from_iterable(protected_met_rxns))))\n",
    "    \n",
    "    # Set constraints so that reactions associated with protected metabolites have at least \n",
    "    # an alpha of 1:\n",
    "    id_met_rxn_protected = []\n",
    "    for rxn in protected_met_rxns:\n",
    "        id_met_rxn_protected.append(rxns.index(rxn))\n",
    "    prot_met_vec = np.zeros(n_rxns)\n",
    "    prot_met_vec[np.array(id_met_rxn_protected)] = 1\n",
    "    \n",
    "    id_rxn_protected = []\n",
    "    for rxn in protected_rxns:\n",
    "        id_rxn_protected.append(rxns.index(rxn))\n",
    "    prot_rxn_vec = np.zeros(n_rxns)\n",
    "    prot_rxn_vec[np.array(id_rxn_protected)] = 1\n",
    "    \n",
    "    # Set upper and lower matrices:\n",
    "    lb_mat = np.array(([lb]*n_functions)).T\n",
    "    ub_mat = np.array(([ub]*n_functions)).T\n",
    "    \n",
    "    # Set bounds according to input df:\n",
    "    for j in range(0,n_functions):\n",
    "        # Get id of reactions in function first column:\n",
    "        id_vec_tmp = [rxns.index(x) for x in functionality_df.iloc[j,0]]\n",
    "        lb_mat[id_vec_tmp,j] = functionality_df.iloc[j,1]\n",
    "        ub_mat[id_vec_tmp,j] = functionality_df.iloc[j,2]\n",
    "    \n",
    "    # Set the steady-state conditions:\n",
    "    b_mat = np.zeros((n_functions,n_mets)).T\n",
    "    \n",
    "    curr_status = 1\n",
    "    k = 0\n",
    "    alpha_list = []\n",
    "    while curr_status == 1:\n",
    "        # Set the inequalities:\n",
    "        alpha_output = cp.Variable(n_rxns, boolean = True)\n",
    "        alpha_orig = cp.Variable((n_rxns,n_functions),boolean = True) # Create a matrix of unknown alphas.\n",
    "        v_vector_orig = cp.Variable((n_rxns,n_functions)) # Create a matrix of unknown fluxes.\n",
    "        my_constraints = []\n",
    "        my_constraints.append(A @ v_vector_orig == b_mat)\n",
    "        my_constraints.append(v_vector_orig >= lb_mat)\n",
    "        my_constraints.append(v_vector_orig <= ub_mat)\n",
    "        my_constraints.append(delta*alpha_orig <= v_vector_orig)\n",
    "        #my_constraints.append(0 <= v_vector_orig)\n",
    "        my_constraints.append(v_vector_orig <= cp.multiply(alpha_orig,M))\n",
    "        #my_constraints.append(v_vector_orig <= alpha_orig*M)\n",
    "\n",
    "        # Add the protected reaction constraints:\n",
    "        my_constraints.append(cp.sum(cp.multiply(prot_rxn_vec,alpha_output)) == len(id_rxn_protected))\n",
    "\n",
    "        # Add the protected metabolite constraints:\n",
    "        my_constraints.append(cp.sum(cp.multiply(prot_met_vec,alpha_output)) >= 1)\n",
    "\n",
    "        # Calculate the sum of alphas:\n",
    "        alpha_sum = cp.sum(alpha_orig, axis=1, keepdims=False)\n",
    "        my_constraints.append(alpha_output <= alpha_sum)\n",
    "        my_constraints.append(alpha_sum <= (1+n_functions)*alpha_output)\n",
    "\n",
    "        # Add the iterative Z constraint:\n",
    "        if k != 0:\n",
    "            #print(cp.sum(cp.multiply(1-Z),alpha_output))\n",
    "            my_constraints.append(cp.sum(cp.multiply((1-np.array(Z)),alpha_output)) + cp.sum(cp.multiply(np.array(Z),(1-alpha_output))) >= 1)\n",
    "            #my_constraints.append(cp.sum(alpha_output) == np.sum(Z))\n",
    "        # Solve!\n",
    "        prob = cp.Problem(cp.Minimize(cp.sum(alpha_output)),constraints=my_constraints)\n",
    "        #prob.solve(solver = 'GLPK_MI',options=[\"--tmlim\", \"10\",\"--interior\"])\n",
    "        prob.solve(solver = 'CPLEX')\n",
    "\n",
    "        if prob.status == 'optimal':\n",
    "            #if k != 0:\n",
    "                #print(np.sum(alpha_output.value-Z))\n",
    "            Z = alpha_output.value.tolist()\n",
    "            if Z in alpha_list:\n",
    "                curr_status = 0\n",
    "                #break\n",
    "            else:\n",
    "                alpha_list.append(Z)\n",
    "                curr_status = 1\n",
    "                k += 1\n",
    "        else:\n",
    "            curr_status = 0\n",
    "            \n",
    "    return alpha_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "77c87087",
   "metadata": {},
   "outputs": [],
   "source": [
    "def new_model_FBA(model_dict,objective_reaction):\n",
    "    import cobra\n",
    "    import cvxpy as cp\n",
    "    import cvxopt\n",
    "    import numpy as np\n",
    "    \n",
    "    A_new = model_dict['A']\n",
    "    lb_new = model_dict['lb']\n",
    "    ub_new = model_dict['ub']\n",
    "    rxns_new = model_dict['rxns']\n",
    "    n_rxns = len(rxns_new)\n",
    "    id_c = rxns_new.index(objective_reaction)\n",
    "    c = np.zeros(A_new.shape[1])\n",
    "    c[id_c] = -1 \n",
    "    lb_obj = model_dict['lb'][id_c]\n",
    "    ub_obj = model_dict['ub'][id_c]\n",
    "    \n",
    "     # Steady-state constraints\n",
    "    b = np.zeros(A_new.shape[0])\n",
    "    \n",
    "    # Create optimization problem:\n",
    "    flux_vector = cp.Variable(n_rxns)\n",
    "    prob = cp.Problem(cp.Minimize(c.T@flux_vector),\n",
    "                 [A_new @ flux_vector == b,\n",
    "                 flux_vector <= ub_new,\n",
    "                 flux_vector >= lb_new,\n",
    "                 c.T@flux_vector <= -lb_obj,\n",
    "                 c.T@flux_vector >= -ub_obj])\n",
    "    #prob.solve(solver = 'GLPK_MI')\n",
    "    prob.solve(solver = 'CPLEX')\n",
    "    if prob.status != 'optimal':\n",
    "            output_vector = [np.nan]*n_rxns\n",
    "            statuss = 'infeasible'\n",
    "            fba_output = pd.Series(dict(zip(rxns_new,output_vector)))\n",
    "    else:\n",
    "            statuss = 'optimal'\n",
    "            fba_output = pd.Series(dict(zip(rxns_new,flux_vector.value)))\n",
    "    \n",
    "    return statuss, fba_output\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e7774a00",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a function that automatically relaxes a model to get feasible flux distributions:\n",
    "def relaxed_FBA(model_dict,obj_function,rxns_to_relax,type_fba = 'normal'):\n",
    "    import cvxpy as cp\n",
    "    import cvxopt\n",
    "    import cobra\n",
    "    # This function relaxes model constraints as little as possible, and outputs a flux distribution\n",
    "    # where at least lb of the objective function is produced. \n",
    "    \n",
    "    \n",
    "    # Define variables\n",
    "    \n",
    "    # S-matrix (as A):\n",
    "    A = model_dict['A']\n",
    "    rxns = model_dict['rxns']\n",
    "    n_rxns = len(rxns)\n",
    "    #rxns = [rxn.id for rxn in model.reactions]\n",
    "    tol = 1e-5\n",
    "    \n",
    "    # Bounds\n",
    "    lb = model_dict['lb']\n",
    "    ub = model_dict['ub']\n",
    "    \n",
    "    \n",
    "    # Objective function vector\n",
    "    id_c = rxns.index(obj_function)\n",
    "    #min_obj = lb[id_c]\n",
    "    c = np.zeros(A.shape[1])\n",
    "    c[id_c] = 1\n",
    "    \n",
    "    # Extract bounds for objective:\n",
    "    lb_obj = lb[id_c]\n",
    "    ub_obj = ub[id_c]\n",
    "    \n",
    "    # Steady-state constraints\n",
    "    b = np.zeros(A.shape[0])\n",
    "\n",
    "    # Start creating problem. Define decision variables:\n",
    "    x_relaxed = cp.Variable(n_rxns)\n",
    "    relax_amount = cp.Variable(n_rxns)\n",
    "    #relax_amount_neg = cp.Variable(n_rxns)\n",
    "    \n",
    "    # Get indeces of relaxation reactions:\n",
    "    if len(rxns_to_relax) < 1:\n",
    "        id_reactions = []\n",
    "    else:\n",
    "        id_reactions = [i for i in range(len(rxns)) if rxns[i] in rxns_to_relax]\n",
    "    \n",
    "    # Create penalization vector:\n",
    "    pen_vec = [1]*n_rxns\n",
    "    id_to_pen = id_reactions\n",
    "    pen_vec = [pen_vec[i] if i in id_to_pen else 1000+pen_vec[i] for i in range(0,len(pen_vec))]\n",
    "    \n",
    "    # Create an empty list to add constraints:\n",
    "    constraints_to_add = []\n",
    "    # Add steady-state constraint:\n",
    "    constraints_to_add.append(A @ x_relaxed == b)\n",
    "    constraints_to_add.append(x_relaxed <= ub+relax_amount)\n",
    "    # Add lower bounds:\n",
    "    constraints_to_add.append(x_relaxed >= lb)\n",
    "    # Add the constraint that relax must be positive:\n",
    "    constraints_to_add.append(relax_amount >= 0)\n",
    "    # Add min and max objective value:\n",
    "    constraints_to_add.append(c.T@x_relaxed >= lb_obj)\n",
    "    constraints_to_add.append(c.T@x_relaxed <= ub_obj)\n",
    "    \n",
    "    \n",
    "    # Create LP problem:\n",
    "    prob = cp.Problem(cp.Minimize(cp.sum(cp.abs(pen_vec@relax_amount))),constraints_to_add)\n",
    "    \n",
    "    # Solve problem:\n",
    "    #prob.solve(solver = 'GLPK_MI')\n",
    "    prob.solve(solver = 'CPLEX')\n",
    "    if type_fba == 'parsimonious':\n",
    "        if prob.status != 'optimal':\n",
    "            output_vector = [np.nan]*n_rxns\n",
    "            #statuss = 'infeasible'\n",
    "            fba_sol = pd.Series(dict(zip(rxns,output_vector)))\n",
    "        else:\n",
    "    ###########################################\n",
    "            print('Relaxation successful - finding minimal solution:')\n",
    "            #print(x_relaxed.value)\n",
    "            # OPTIMIZATION PROBLEM 2:\n",
    "            # Create a new FBA problem, where the total flux is minimized keeping the same obj value:\n",
    "            new_upper_bound = ub + relax_amount.value\n",
    "            obj_val = abs(x_relaxed.value[id_c])\n",
    "            #print(obj_val)\n",
    "            pfba_flux_vector = cp.Variable(n_rxns)\n",
    "            prob2 = cp.Problem(cp.Minimize(cp.sum(pfba_flux_vector)),\n",
    "                         [A @ pfba_flux_vector == b,\n",
    "                          #cp.abs(cp.sum(b)) <= tol,\n",
    "                         pfba_flux_vector <= new_upper_bound,\n",
    "                         pfba_flux_vector >= lb,\n",
    "                         #c.T@pfba_flux_vector >= 0.99*obj_val,\n",
    "                         #c.T@pfba_flux_vector <= obj_val],\n",
    "                         c.T@pfba_flux_vector == obj_val])\n",
    "            #prob2.solve(solver = 'GLPK_MI')\n",
    "            prob2.solve(solver = 'CPLEX')\n",
    "            print('pFBA solving status:',prob2.status)\n",
    "            if prob2.status != 'optimal':\n",
    "                output_vector = [np.nan]*n_rxns\n",
    "                #statuss = 'infeasible'\n",
    "                fba_sol = pd.Series(dict(zip(rxns,output_vector)))\n",
    "            else:\n",
    "                output_vector = pfba_flux_vector.value\n",
    "                print(output_vector)\n",
    "                output_vector[abs(output_vector) < 1e-6] = 0\n",
    "                fba_sol = pd.Series(dict(zip(rxns,output_vector)))\n",
    "                # Print result.\n",
    "                #print(relax_amount.value)\n",
    "                id_nonzero = np.where(abs(relax_amount.value) >= tol)[0].tolist()\n",
    "                rel_rxns = [rxns[i] for i in id_nonzero]\n",
    "                rel_amount = [relax_amount.value[i] for i in id_nonzero]\n",
    "                orig_ub = ub\n",
    "                diff_amount = [x2-x1 for (x1, x2) in zip(orig_ub, rel_amount)]\n",
    "                rel_dict = dict(zip(rel_rxns,diff_amount))\n",
    "                print('\\nRelaxation reactions: ')\n",
    "                for keys,values in rel_dict.items():\n",
    "                    print(keys, round(values,2),'relaxation in upper bound')\n",
    "                    #print(values)\n",
    "                #if problem.status not in [\"infeasible\", \"unbounded\"]:\n",
    "                # Otherwise, problem.value is inf or -inf, respectively.\n",
    "                #print(\"Optimal value: %s\" % problem.value)\n",
    "                print(\"\\nTotal amount of relaxation: \",prob.value)\n",
    "    elif type_fba == 'normal':\n",
    "        if prob.status != 'optimal':\n",
    "            output_vector = [np.nan]*n_rxns\n",
    "            #statuss = 'infeasible'\n",
    "            fba_sol = pd.Series(dict(zip(rxns,output_vector)))\n",
    "            new_upper_bound = [np.nan]*n_rxns\n",
    "        else:\n",
    "            fba_sol = pd.Series(dict(zip(rxns,x_relaxed.value)))\n",
    "            id_nonzero = np.where(abs(relax_amount.value) >= tol)[0].tolist()\n",
    "            rel_rxns = [rxns[i] for i in id_nonzero]\n",
    "            rel_amount = [relax_amount.value[i] for i in id_nonzero]\n",
    "            orig_ub =  ub\n",
    "            diff_amount = [x2-x1 for (x1, x2) in zip(orig_ub, rel_amount)]\n",
    "            rel_dict = dict(zip(rel_rxns,diff_amount))\n",
    "            print('\\nRelaxation reactions: ')\n",
    "            for keys,values in rel_dict.items():\n",
    "                print(keys, round(values,2),'relaxation in upper bound')\n",
    "            print(\"\\nTotal amount of relaxation: \",prob.value)\n",
    "            new_upper_bound = ub + relax_amount.value \n",
    "            new_upper_bound[new_upper_bound < tol] = 0\n",
    "    return fba_sol, new_upper_bound"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0a2063db",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reduced_model_network(model_orig,new_alpha):\n",
    "    import cobra\n",
    "    import numpy as np\n",
    "    A = cobra.util.array.create_stoichiometric_matrix(model,'dense')\n",
    "    rxns = [rxn.id for rxn in model.reactions]\n",
    "    mets = [met.id for met in model.metabolites]\n",
    "    lb = [rxn.lower_bound for rxn in model.reactions]\n",
    "    ub = [rxn.upper_bound for rxn in model.reactions]\n",
    "    \n",
    "    # Output a new model (S-matrix):\n",
    "    id_present_reactions = np.where(new_alpha)\n",
    "    A2 = A.T[id_present_reactions].T\n",
    "    id_metabolites_present = np.where(A2.any(axis=1))[0]\n",
    "    A_new = np.delete(A2, np.where(~A2.any(axis=1))[0], 0)\n",
    "    \n",
    "    # Output new lbs, ubs, rxns and metabolites for the model:\n",
    "    lb_new = list(np.array(lb)[id_present_reactions])\n",
    "    ub_new = list(np.array(ub)[id_present_reactions])\n",
    "    rxns_new = list(np.array(rxns)[id_present_reactions])\n",
    "    metabolites = [met.id for met in model.metabolites]\n",
    "    metabolites_new = list(np.array(metabolites)[id_metabolites_present])\n",
    "    \n",
    "    # Output a model as dictionary:\n",
    "    model_dict = defaultdict()\n",
    "    model_dict['A'] = A_new\n",
    "    model_dict['rxns'] = rxns_new\n",
    "    model_dict['mets'] = metabolites_new\n",
    "    model_dict['ub'] = ub_new\n",
    "    model_dict['lb'] = lb_new\n",
    "    return model_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "870d5c62",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_futile_cycles(A_new,efms):\n",
    "    '''\n",
    "    A function to remove futile cycles from an array of EFMs (e.g. calculated from EFMtool)\n",
    "    :param A_new: A Stoichiometric matrix in array format\n",
    "    :param efms: An array of efms\n",
    "    :return: efm_out: An array where thermodynamically unwanted efms are removed\n",
    "    '''\n",
    "    import numpy as np\n",
    "\n",
    "    id_to_remove = []\n",
    "    for ele in range(0,efms.shape[1]):\n",
    "        efm = efms.T[ele]\n",
    "        location_reactions = np.where(efm)[0]\n",
    "        if len(location_reactions) == 2:\n",
    "            id_1 = location_reactions[0]\n",
    "            id_2 = location_reactions[1]\n",
    "            if all(np.where(A_new.T[id_1])[0] == np.where(A_new.T[id_2])[0]):\n",
    "                # Found a futile cycle - dont do anything.\n",
    "                #id_to_remove.append(ele)\n",
    "                continue\n",
    "        else:\n",
    "            id_to_remove.append(ele)\n",
    "    efms_out = efms.T[np.array(id_to_remove)].T\n",
    "    #print(id_to_remove)\n",
    "    return efms_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7e3e4039",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_functionality_feasibility(model,functionality_df):\n",
    "    '''\n",
    "    A function to check whether all provided functionalities in a functionality df are feasible, i.e. an FBA is run for all\n",
    "    of them. If the functionality is not feasible, the bounds are relaxed so at least the lb of the biomass is produced.\n",
    "    '''\n",
    "    feasible = []\n",
    "    mdl = model.copy()\n",
    "    for func_id in range(0,functionality_df.shape[0]):\n",
    "        # Constrain model according to provided feasibilities:\n",
    "        for idxx,rxn in enumerate(functionality_df.iloc[func_id,0]):\n",
    "            mdl.reactions.get_by_id(rxn).bounds = (functionality_df.iloc[func_id,1][idxx],functionality_df.iloc[func_id,2][idxx])\n",
    "        if not np.isnan(mdl.slim_optimize()):\n",
    "            feasible.append('feasible')\n",
    "        else:\n",
    "            feasible.append('infeasible')\n",
    "    return feasible\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "986f0b64",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_mets_from_rxns(model, rxn_list):\n",
    "    import cobra\n",
    "    import numpy as np\n",
    "\n",
    "    if type(rxn_list) != list:\n",
    "        rxn_list = [rxn_list]\n",
    "    A = cobra.util.array.create_stoichiometric_matrix(model, 'dense')\n",
    "    rxns = [rxn.id for rxn in model.reactions]\n",
    "    mets = [met.id for met in model.metabolites]\n",
    "    if len(rxn_list) == 1:\n",
    "        tmp_rxn = rxn_list[0]\n",
    "        id_rxn = rxns.index(tmp_rxn)\n",
    "        outp = list(np.array(mets)[np.where(A[:, id_rxn])])\n",
    "    elif len(rxn_list) == 0:\n",
    "        print('Put in a reaction id!')\n",
    "    else:\n",
    "        # Find mets that belong to all rxns:\n",
    "        outp = []\n",
    "        met_list = []\n",
    "        for tmp_rxn in rxn_list:\n",
    "            id_rxn = rxns.index(tmp_rxn)\n",
    "            met_list.append(list(np.array(mets)[np.where(A[:, id_rxn])]))\n",
    "        list_to_compare = met_list[0]\n",
    "        for met_tmp in list_to_compare:\n",
    "            count = 1\n",
    "            for i in range(1, len(met_list)):\n",
    "                if met_tmp in met_list[i]:\n",
    "                    count += 1\n",
    "            if count == len(met_list):\n",
    "                outp.append(met_tmp)\n",
    "    return outp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b46c6de2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_rxns_from_mets(model,met_list):\n",
    "    import cobra\n",
    "    import numpy as np\n",
    "\n",
    "    if type(met_list) != list:\n",
    "        met_list = [met_list]\n",
    "    A = cobra.util.array.create_stoichiometric_matrix(model,'dense')\n",
    "    rxns = [rxn.id for rxn in model.reactions]\n",
    "    mets = [met.id for met in model.metabolites]\n",
    "    if len(met_list) == 1:\n",
    "        tmp_met = met_list[0]\n",
    "        id_met = mets.index(tmp_met)\n",
    "        outp = list(np.array(rxns)[np.where(A[id_met,:])])\n",
    "    elif len(met_list) == 0:\n",
    "        print('Put in a metabolite id!')\n",
    "    else:\n",
    "        # Find rxns that contain all metabolites:\n",
    "        outp = []\n",
    "        rxn_list = []\n",
    "        for tmp_met in met_list:\n",
    "            id_met = mets.index(tmp_met)\n",
    "            rxn_list.append(list(np.array(rxns)[np.where(A[id_met,:])]))\n",
    "        list_to_compare = rxn_list[0]\n",
    "        for rxn_tmp in list_to_compare:\n",
    "            count = 1\n",
    "            for i in range(1,len(rxn_list)):\n",
    "                if rxn_tmp in rxn_list[i]:\n",
    "                    count += 1\n",
    "            if count == len(rxn_list):\n",
    "                outp.append(rxn_tmp)\n",
    "    return outp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1f2df482",
   "metadata": {},
   "outputs": [],
   "source": [
    "def indices(lst, element):\n",
    "    result = []\n",
    "    offset = -1\n",
    "    while True:\n",
    "        try:\n",
    "            offset = lst.index(element, offset+1)\n",
    "        except ValueError:\n",
    "            return result\n",
    "        result.append(offset)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4866acb",
   "metadata": {},
   "source": [
    "### Load E.coli model, prepare functionality df and run the MinNW algorithm:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b7dfdec4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Restricted license - for non-production use only - expires 2023-10-25\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>e_coli_core</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07fd7d00b8580</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>72</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>141</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*Biomass_Ecoli_core - 1.0*Biomass_Ecoli_core_reverse_2cdba</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model e_coli_core at 0x7fd7d00b8580>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = cobra.test.create_test_model(\"textbook\")\n",
    "convert_to_irreversible(model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "57c7eaec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "E.coli model has dimensions:  (72, 141)\n"
     ]
    }
   ],
   "source": [
    "# Get info on model size:\n",
    "A0 = cobra.util.array.create_stoichiometric_matrix(model,'dense')\n",
    "print('E.coli model has dimensions: ',A0.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c89903e",
   "metadata": {},
   "source": [
    "#### A) Try out the algorithm using two simple functionalities (aerobic and anaerobic biomass production):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "35d942b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "funcs = [['Biomass_Ecoli_core'],['Biomass_Ecoli_core','EX_o2_e_b']]\n",
    "lbs = [[0.99*0.87],[0.99*0.21,0]]\n",
    "ubs = [[1000],[1000,0]]\n",
    "func_df_ecoli = pd.DataFrame([funcs,lbs,ubs]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a9b5d882",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /tmp/tmpo0fot24g.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 72 rows, 282 columns, 1026 nonzeros\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['feasible', 'feasible']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if all functionalities result in a feasible FBA solution:\n",
    "check_functionality_feasibility(model,func_df_ecoli)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6a08ae0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "optimal\n",
      "CPU times: user 2.1 s, sys: 1.42 s, total: 3.52 s\n",
      "Wall time: 316 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "protected_mets = find_mets_from_rxns(model,'Biomass_Ecoli_core')\n",
    "stat_ecoli,model_dict_ecoli = MinNW(model,protected_mets,['Biomass_Ecoli_core'],func_df_ecoli)\n",
    "print(stat_ecoli)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c041dfe6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "optimal\n",
      "0.8638133095040007\n"
     ]
    }
   ],
   "source": [
    "# Check the optimal value when biomass is optimized:\n",
    "stat2_ecoli,fba_ecoli = new_model_FBA(model_dict_ecoli,'Biomass_Ecoli_core')\n",
    "print(stat2_ecoli)\n",
    "print(fba_ecoli['Biomass_Ecoli_core'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c88e3fc9",
   "metadata": {},
   "source": [
    "#### B) Try out the algorithm using a more complex combination of functionalities:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0cf6856f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform the original FBA to get the bounds:\n",
    "new_bounds = fba_ecoli*0.9\n",
    "\n",
    "# How many random values do I want?\n",
    "X = 3\n",
    "\n",
    "# Get the ids of the reactions I want to change:\n",
    "rxns = ['Biomass_Ecoli_core','EX_ac_e','EX_co2_e_f','EX_etoh_e','EX_for_e','EX_glc__D_e_b','EX_o2_e_b']\n",
    "rxns_fin = [rxns]*X\n",
    "\n",
    "# Create X random values between 0.5 and 1:\n",
    "random_numbers = np.random.random((X,len(rxns)))\n",
    "\n",
    "# Create index list of occuring reactions:\n",
    "indices_list = []\n",
    "for rxn in rxns:\n",
    "    indices_list.append(indices(model_dict_ecoli['rxns'],rxn)[0])\n",
    "\n",
    "\n",
    "lbs = np.array([abs(np.array(new_bounds))[np.array(indices_list)]*fraction for fraction in random_numbers])\n",
    "lbs = [list(lbs[i,:]) for i in range(lbs.shape[0])]\n",
    "\n",
    "ubs = np.array([np.array(model_dict_ecoli['ub'])[np.array(indices_list)*1] for fraction in random_numbers])\n",
    "ubs = [list(ubs[i,:]) for i in range(ubs.shape[0])]\n",
    "\n",
    "func_df_ecoli_large = pd.DataFrame([rxns_fin,lbs,ubs]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6a5ea3af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /tmp/tmppwpyow7h.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 72 rows, 282 columns, 1026 nonzeros\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['feasible', 'feasible', 'feasible']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_functionality_feasibility(model,func_df_ecoli_large)\n",
    "# If any of the functionalities display \"infeasible\" here, something is wrong..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c83eea7e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "optimal\n",
      "CPU times: user 3.78 s, sys: 66.6 ms, total: 3.85 s\n",
      "Wall time: 321 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "protected_mets = find_mets_from_rxns(model,'Biomass_Ecoli_core')\n",
    "stat_ecoli_large,model_dict_ecoli_large = MinNW(model,protected_mets,['Biomass_Ecoli_core'],func_df_ecoli_large)\n",
    "print(stat_ecoli_large)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fbb48a1",
   "metadata": {},
   "source": [
    "### Run the elementary flux mode (EFM) enumeration:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "60fa3c67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reduced E.coli model EFMs:  37\n"
     ]
    }
   ],
   "source": [
    "rev = [0]*model_dict_ecoli['A'].shape[1]\n",
    "reactions = model_dict_ecoli['rxns']\n",
    "metabolites = model_dict_ecoli['mets']\n",
    "efms_ecoli_red = efmtool.calculate_efms(model_dict_ecoli['A'],rev,reactions,metabolites)\n",
    "efms_ecoli_red = remove_futile_cycles(model_dict_ecoli['A'],efms_ecoli_red)\n",
    "print('Number of reduced E.coli model EFMs: ',efms_ecoli_red.shape[1])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
