{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9208bb63",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.mode.chained_assignment = None\n",
    "import copy\n",
    "import cobra\n",
    "import cobra.test\n",
    "import math\n",
    "from collections import defaultdict\n",
    "#import efmtool\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e4ba4a64",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cvxpy as cp\n",
    "import cvxopt\n",
    "import cobra\n",
    "import itertools\n",
    "from cobra.flux_analysis import flux_variability_analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e479650",
   "metadata": {},
   "source": [
    "## Define functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d006a460",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_rxns_from_mets(model,met_list):\n",
    "    import cobra\n",
    "    import numpy as np\n",
    "\n",
    "    if type(met_list) != list:\n",
    "        met_list = [met_list]\n",
    "    A = cobra.util.array.create_stoichiometric_matrix(model,'dense')\n",
    "    rxns = [rxn.id for rxn in model.reactions]\n",
    "    mets = [met.id for met in model.metabolites]\n",
    "    if len(met_list) == 1:\n",
    "        tmp_met = met_list[0]\n",
    "        id_met = mets.index(tmp_met)\n",
    "        outp = list(np.array(rxns)[np.where(A[id_met,:])])\n",
    "    elif len(met_list) == 0:\n",
    "        print('Put in a metabolite id!')\n",
    "    else:\n",
    "        # Find rxns that contain all metabolites:\n",
    "        outp = []\n",
    "        rxn_list = []\n",
    "        for tmp_met in met_list:\n",
    "            id_met = mets.index(tmp_met)\n",
    "            rxn_list.append(list(np.array(rxns)[np.where(A[id_met,:])]))\n",
    "        list_to_compare = rxn_list[0]\n",
    "        for rxn_tmp in list_to_compare:\n",
    "            count = 1\n",
    "            for i in range(1,len(rxn_list)):\n",
    "                if rxn_tmp in rxn_list[i]:\n",
    "                    count += 1\n",
    "            if count == len(rxn_list):\n",
    "                outp.append(rxn_tmp)\n",
    "    return outp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6d4bba50",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_to_irreversible(cobra_model):\n",
    "    \"\"\"\n",
    "    Split reversible reactions into two irreversible reactions\n",
    "\n",
    "    These two reactions will proceed in opposite directions. This\n",
    "    guarentees that all reactions in the model will only allow\n",
    "    positive flux values, which is useful for some modeling problems.\n",
    "\n",
    "    cobra_model:\n",
    "        A cobra model object which will be modified in place.\n",
    "\n",
    "    \"\"\"\n",
    "    from cobra import Reaction\n",
    "    import pandas as pd\n",
    "    from cobra.util import set_objective\n",
    "    reactions_to_add = []\n",
    "    coefficients = {}\n",
    "    for reaction in cobra_model.reactions:\n",
    "        # If a reaction is reverse only, the forward reaction (which\n",
    "        # will be constrained to 0) will be left in the model.\n",
    "        if reaction.lower_bound < 0:\n",
    "            reverse_reaction = Reaction(reaction.id + \"_b\")\n",
    "            reaction.id = reaction.id + \"_f\"\n",
    "            reverse_reaction.lower_bound = max(0, -reaction.upper_bound)\n",
    "            reverse_reaction.upper_bound = -reaction.lower_bound\n",
    "            coefficients[reverse_reaction] = reaction.objective_coefficient * -1\n",
    "            reaction.upper_bound = max(0, reaction.upper_bound)\n",
    "            reaction.lower_bound = max(0, reaction.lower_bound)\n",
    "\n",
    "            # Make the directions aware of each other\n",
    "            reaction.notes[\"reflection\"] = reverse_reaction.id\n",
    "            reverse_reaction.notes[\"reflection\"] = reaction.id\n",
    "            reaction_dict = {k: v * -1 for k, v in pd.Series.iteritems(reaction._metabolites)}\n",
    "            reverse_reaction.add_metabolites(reaction_dict)\n",
    "            reverse_reaction._model = reaction._model\n",
    "            reverse_reaction._genes = reaction._genes\n",
    "            for gene in reaction._genes:\n",
    "                gene._reaction.add(reverse_reaction)\n",
    "            reverse_reaction.subsystem = reaction.subsystem\n",
    "            #reverse_reaction._gene_reaction_rule = reaction._gene_reaction_rule\n",
    "            #\n",
    "            # ALO edited previous line as\n",
    "            #\n",
    "            reverse_reaction._gene_reaction_rule = reaction.gene_reaction_rule\n",
    "            reactions_to_add.append(reverse_reaction)\n",
    "    cobra_model.add_reactions(reactions_to_add)\n",
    "    set_objective(cobra_model, coefficients, additive=True)\n",
    "    return cobra_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "99a880cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def MinNW(model, protected_mets, protected_rxns, functionality_df, selected_solver):\n",
    "    '''\n",
    "    A MILP formulation derived from (https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-016-1412-z) to\n",
    "    reduce a Genome-Scale Metabolic Network so that EFM calculation is faster and macroreactions may be derived. This is \n",
    "    an approach that InSilico uses. I want to make sure that it works for my stuff as well. \n",
    "    This is the second MILP, MinNW, that finds the smallest network it can given a model and specific functionalities. \n",
    "    \n",
    "    Inputs:\n",
    "        model: A cobra genome-scale metabolic model\n",
    "        protected_mets: A list of protected metabolites, i.e. metabolites that absolutely need to be in the resulting network\n",
    "        \n",
    "        protected_rxns: A list of protected reactions, i.e. reactions that absolutely need to be in the resulting network\n",
    "        \n",
    "        functionality_df: A (Fx3) pandas DataFrame containing the reactions along with their lower and upper bounds describing \n",
    "            the F functionalities desired in the resulting network. May be opposing ones (e.g. O2 uptake inhibited and allowed).\n",
    "            The first column contains the name of the reactions of interest, the second contains the lower bound and third\n",
    "            the upper bound for the reaction. May contain more than one reaction. Needs to contain an equivalent amount of bonds. \n",
    "        \n",
    "    Outputs:\n",
    "        prob.status: Status of solver, either 'infeasible' or 'optimal'\n",
    "        \n",
    "        model_dict: A dictionary containing the Stoichiometric matrix, reactions, metabolites and bounds for the new, reduced \n",
    "            network. \n",
    "    \n",
    "    '''\n",
    "    #import cvxpy as cp\n",
    "    #import cvxopt\n",
    "    #import cobra\n",
    "    #import itertools\n",
    "    #from cobra.flux_analysis import flux_variability_analysis\n",
    "    \n",
    "    #\n",
    "    # ALO edited. Do not do imports inside a function, specially if you time it. imports should be at the top of a file always\n",
    "    #\n",
    "    \n",
    "    A = cobra.util.array.create_stoichiometric_matrix(model,'dense')\n",
    "    b = np.zeros(A.shape[0])\n",
    "    rxns = [rxn.id for rxn in model.reactions]\n",
    "    n_rxns = len(rxns)\n",
    "    n_mets = A.shape[0]\n",
    "    n_functions = functionality_df.shape[0]\n",
    "    lb = [rxn.lower_bound for rxn in model.reactions]\n",
    "    ub = [rxn.upper_bound for rxn in model.reactions]\n",
    "    #M = max(ub)\n",
    "    \n",
    "    # Perform FVA to get correct M:\n",
    "    #fva_df = flux_variability_analysis(model,fraction_of_optimum=0.99) # TODO: Get lowest objective value from functionality here..\n",
    "    #new_ub = fva_df['maximum'].tolist()\n",
    "    #M = np.array([new_ub]).T\n",
    "    M = np.array([ub]).T\n",
    "    #print(M)\n",
    "    #print(M.shape)\n",
    "    delta = 1e-6 # Between 1e-4 and 1e-6\n",
    "    #delta = np.array([lb]).T # Between 1e-4 and 1e-6\n",
    "    \n",
    "    # Initiate an id_protected_rxns list:\n",
    "    protected_rxns_orig = list(itertools.chain.from_iterable(protected_rxns))\n",
    "    protected_met_rxns = []\n",
    "    for met in protected_mets:\n",
    "        protected_met_rxns_tmp = find_rxns_from_mets(model,met)\n",
    "        if bool(set(protected_rxns_orig) & set(protected_met_rxns_tmp)):\n",
    "            continue\n",
    "        else:\n",
    "            protected_met_rxns.append(protected_met_rxns_tmp)\n",
    "    protected_met_rxns = list(set(list(itertools.chain.from_iterable(protected_met_rxns))))\n",
    "    \n",
    "    # Set constraints so that reactions associated with protected metabolites have at least \n",
    "    # an alpha of 1:\n",
    "    id_met_rxn_protected = []\n",
    "    for rxn in protected_met_rxns:\n",
    "        id_met_rxn_protected.append(rxns.index(rxn))\n",
    "    prot_met_vec = np.zeros(n_rxns)\n",
    "    prot_met_vec[np.array(id_met_rxn_protected)] = 1\n",
    "    \n",
    "    id_rxn_protected = []\n",
    "    for rxn in protected_rxns:\n",
    "        id_rxn_protected.append(rxns.index(rxn))\n",
    "    prot_rxn_vec = np.zeros(n_rxns)\n",
    "    prot_rxn_vec[np.array(id_rxn_protected)] = 1\n",
    "    \n",
    "    # Add constraints to a matrix instead of multiple dictionaries:\n",
    "    alpha_output = cp.Variable(n_rxns,boolean=True)\n",
    "    alpha_orig = cp.Variable((n_rxns,n_functions),boolean=True) # Create a matrix of unknown alphas.\n",
    "    v_vector_orig = cp.Variable((n_rxns,n_functions)) # Create a matrix of unknown fluxes.\n",
    "    \n",
    "    # Set upper and lower matrices:\n",
    "    lb_mat = np.array(([lb]*n_functions)).T\n",
    "    ub_mat = np.array(([ub]*n_functions)).T\n",
    "    \n",
    "    # Set bounds according to input df:\n",
    "    for j in range(0,n_functions):\n",
    "        # Get id of reactions in function first column:\n",
    "        id_vec_tmp = [rxns.index(x) for x in functionality_df.iloc[j,0]]\n",
    "        lb_mat[id_vec_tmp,j] = functionality_df.iloc[j,1]\n",
    "        ub_mat[id_vec_tmp,j] = functionality_df.iloc[j,2]\n",
    "    \n",
    "    # Set the steady-state conditions:\n",
    "    b_mat = np.zeros((n_functions,n_mets)).T\n",
    "    \n",
    "    # Set the inequalities:\n",
    "    my_constraints = []\n",
    "    my_constraints.append(A @ v_vector_orig == b_mat)\n",
    "    my_constraints.append(v_vector_orig >= lb_mat)\n",
    "    my_constraints.append(v_vector_orig <= ub_mat)\n",
    "    my_constraints.append(delta*alpha_orig <= v_vector_orig)\n",
    "    #my_constraints.append(0 <= v_vector_orig)\n",
    "    my_constraints.append(v_vector_orig <= cp.multiply(alpha_orig,M))\n",
    "    #my_constraints.append(v_vector_orig <= alpha_orig*M)\n",
    "    \n",
    "    # Add the protected reaction constraints:\n",
    "    my_constraints.append(cp.sum(cp.multiply(prot_rxn_vec,alpha_output)) == len(id_rxn_protected))\n",
    "    \n",
    "    # Add the protected metabolite constraints:\n",
    "    my_constraints.append(cp.sum(cp.multiply(prot_met_vec,alpha_output)) >= 1)\n",
    "    \n",
    "    # Calculate the sum of alphas:\n",
    "    alpha_sum = cp.sum(alpha_orig, axis=1, keepdims=False)\n",
    "    my_constraints.append(alpha_output <= alpha_sum)\n",
    "    my_constraints.append(alpha_sum <= (1+n_functions)*alpha_output)\n",
    "    \n",
    "    # Solve!\n",
    "    prob = cp.Problem(cp.Minimize(cp.sum(alpha_output)),constraints=my_constraints)\n",
    "    prob.solve(solver = selected_solver) # TODO: CHANGE TO CPLEX!!!!\n",
    "    \n",
    "    # Output a new model (S-matrix):\n",
    "    id_present_reactions = np.where(alpha_output.value)\n",
    "    A2 = A.T[id_present_reactions].T\n",
    "    id_metabolites_present = np.where(A2.any(axis=1))[0]\n",
    "    A_new = np.delete(A2, np.where(~A2.any(axis=1))[0], 0)\n",
    "    \n",
    "    # Output new lbs, ubs, rxns and metabolites for the model:\n",
    "    lb_new = list(np.array(lb)[id_present_reactions])\n",
    "    ub_new = list(np.array(ub)[id_present_reactions])\n",
    "    rxns_new = list(np.array(rxns)[id_present_reactions])\n",
    "    metabolites = [met.id for met in model.metabolites]\n",
    "    metabolites_new = list(np.array(metabolites)[id_metabolites_present])\n",
    "    \n",
    "    # Output a model as dictionary:\n",
    "    model_dict = defaultdict()\n",
    "    model_dict['A'] = A_new\n",
    "    model_dict['rxns'] = rxns_new\n",
    "    model_dict['mets'] = metabolites_new\n",
    "    model_dict['ub'] = ub_new\n",
    "    model_dict['lb'] = lb_new\n",
    "    \n",
    "    return prob.status,model_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59162cb2",
   "metadata": {},
   "source": [
    "## Load model:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "786fd9bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "E.coli model has dimensions:  (72, 141)\n"
     ]
    }
   ],
   "source": [
    "model = cobra.test.create_test_model(\"textbook\")\n",
    "convert_to_irreversible(model)\n",
    "A0 = cobra.util.array.create_stoichiometric_matrix(model,'dense')\n",
    "print('E.coli model has dimensions: ',A0.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c06dd77c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the functionality for E.coli:\n",
    "funcs = [['Biomass_Ecoli_core']]\n",
    "lbs = [[0.99*0.87]]\n",
    "ubs = [[1000]]\n",
    "func_df_ecoli = pd.DataFrame([funcs,lbs,ubs]).T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98d66a67",
   "metadata": {},
   "source": [
    "## Perform the Network reduction for E.coli:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e1529bb9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "optimal\n",
      "CPU times: user 3.98 s, sys: 0 ns, total: 3.98 s\n",
      "Wall time: 3.97 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "selected_solver = 'GLPK_MI'\n",
    "stat_ecoli,model_dict_ecoli = MinNW(model, ['atp_c'], ['Biomass_Ecoli_core'], func_df_ecoli, selected_solver)\n",
    "print(stat_ecoli)\n",
    "\n",
    "# You could compare the \"GLPK_MI\" and the \"CPLEX\" solvers in terms of time. You could also load the larger\n",
    "# E.coli model by using \"ecoli\" instead of \"textbook\" and see there how the performance of the algorithm is affected\n",
    "# by model size ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "196edca2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "optimal\n",
      "CPU times: user 524 ms, sys: 14.2 ms, total: 538 ms\n",
      "Wall time: 79.9 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "selected_solver = 'CPLEX'\n",
    "stat_ecoli,model_dict_ecoli = MinNW(model, ['atp_c'], ['Biomass_Ecoli_core'], func_df_ecoli, selected_solver)\n",
    "print(stat_ecoli)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
